### PR TITLE
feat(engine): buildInsightSQL filter clause — WHERE/HAVING routing + between operator

### PR DESCRIPTION
**Pull request by youhaowei**
`buildInsightSQL` accepted `Insight.filters` but emitted **no filter clause** — saved filters were silently dropped at query time. This is a correctness bug. This PR wires filter-clause generation into the SQL builder and routes predicates by dimension-vs-metric.

> No public tracking issue exists for this fix (issue #74 is the native-DuckDB engine, unrelated). Tracked internally.

## Changes

- **Generate filter clauses** in `buildInsightSQL` from the effective `Insight.filters` — they were accepted but never emitted before.
- **Route WHERE vs HAVING by dimension-vs-metric, derived at compile time** from the insight definition: a filter whose `field` matches a `selectedFields` dimension → `WHERE` (pre-aggregation); a filter whose field matches a metric column → `HAVING` (post-aggregation, referencing the aggregate expression). No discriminator field was added to `InsightFilter` — array membership in the insight *is* the dimension-vs-metric identity, so there is no desync surface.
- **`between` operator**: inclusive `field BETWEEN low AND high`, value shape `{ low, high }`. A date range is a `between` on a date field, not a separate type. (Type-layer change: `"between"` added to the operator union + new `InsightFilterBetweenValue` interface, exported.)
- **All operators handled**: `eq/ne/gt/gte/lt/lte` (scalar), `contains` (`LIKE %..%` with wildcard escaping), `in` (`IN (...)`), `between`.
- **Injection guard**: values are escaped via SQL single-quote doubling (matching the module's existing inline-literal convention) — no raw string interpolation of values.

## Screenshots

No UI change — backend SQL builder only.

## Verification

- New executable-spec tests in `packages/engine/src/sql/insight-sql.test.ts`, each named for its contract: grouped-dimension filter → WHERE; metric filter → HAVING; mixed → both correctly split; `between` on a date → inclusive `BETWEEN low AND high`, quoted; `in` → `IN (...)`; each scalar operator → correct SQL operator; **regression: an insight WITH saved filters now emits a clause** (asserts the clause is present — today it is silently absent); a value containing a quote char does not break the SQL (injection guard).
- `bun test` in `packages/engine`: **11 pass / 0 fail** (2 pre-existing + 9 new).
- Verified the generated SQL by hand for a mixed filter: `… WHERE "field_<region>" = 'EMEA' GROUP BY "field_<region>" HAVING SUM("field_<amount>") > 1000` — WHERE precedes GROUP BY, HAVING follows it and references the aggregate.
- Full CI-equivalent locally green: `bun format:check`, `bun run check:ticket-refs`, `bun check` (lint + typecheck + test across all 77 tasks).

<details open><summary><h3>Greptile Summary</h3></summary>

This PR wires filter-clause generation into `buildInsightSQL`, fixing a correctness bug where `Insight.filters` were accepted but never emitted. Filters are now routed to WHERE (dimension/pre-aggregation) or HAVING (metric/post-aggregation) based on the insight definition, with the `between` operator added and an injection guard via SQL single-quote doubling.

- **Filter routing**: `isMetricFilter` checks `selectedFields` membership first (dimension wins over metric for shared columns), then metric array membership; `buildFilterClauses` builds WHERE and HAVING predicate lists accordingly and joins them with AND.
- **Operator coverage**: `eq`/`ne` emit IS NULL / IS NOT NULL for null values; `between` guards its value shape and emits `1=1` with a warning on malformed input; `contains` uses `LIKE '%…%' ESCAPE '\\'` with wildcard-char escaping; `in` handles empty arrays as `1=0`.
- **Model-mode consistency**: Both the simple and joined code paths intentionally suppress filters in model mode (raw preview rows), documented in code comments and a dedicated test.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; the filter routing, operator predicates, null guards, and injection escaping are all correct, and the test suite covers all code paths including edge cases.

The implementation is thorough: dimension-vs-metric routing is correctly derived at compile time, IS NULL / IS NOT NULL guards are in place for eq/ne, between has a malformed-value guard, and a fail-safe skip prevents broken column references from crashing queries. The only gap is that gt/gte/lt/lte with a null value silently produce col > NULL (zero rows, no warning) — an edge case inconsistency with the null handling already present for eq/ne and between, but not a regression over previous behavior.

packages/engine/src/sql/insight-sql.ts — specifically the comparison-operator cases in buildFilterPredicate; all other files are clean.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/engine/src/sql/insight-sql.ts | Core implementation of filter clause generation: WHERE/HAVING routing, operator predicates, quoting, and null guards. Logic is well-structured with correct IS NULL handling for eq/ne and a malformed-value guard for between; gt/gte/lt/lte with a null value silently produce col > NULL (returns no rows, no warning). |
| packages/engine/src/sql/insight-sql.test.ts | New test file with 11 specs covering WHERE/HAVING routing, all operators, null predicates, injection guard, model-mode suppression, and join-key drop safety. All critical contracts are covered; no null-value tests for gt/gte/lt/lte operators. |
| packages/types/src/insights.ts | Adds `between` to the operator union and exports new InsightFilterBetweenValue interface; clean, minimal type-layer change. |
| packages/types/src/index.ts | Exports the new InsightFilterBetweenValue type; one-line addition, no issues. |

</details>

</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[buildInsightSQL] --> B{Has joins?}
    B -- Yes --> C[buildJoinedSQL\nreturns sql + availableFields]
    C --> D{mode === model?}
    D -- Yes --> E[SELECT * FROM joined.sql\nno filters]
    D -- No --> F[buildAggregatedSQL\njoined.sql]
    B -- No --> G[buildSimpleSQL]
    G --> H{mode === model OR\nno config?}
    H -- Yes + model --> I[SELECT * FROM table\nno filters]
    H -- Yes + query --> J[SELECT * FROM table\nWHERE buildFilterClauses\nraw refMode]
    H -- No --> K[Wrap table as subquery\nbuildAggregatedSQL]
    F --> L[buildFilterClauses\nalias refMode]
    K --> L
    L --> M{isMetricFilter?}
    M -- Yes + hasAggregation --> N[HAVING agg_expr op val]
    M -- No --> O{resolveFilterColumnRef\nreturns null?}
    O -- null --> P[Skip filter + warn]
    O -- ref --> Q[WHERE ref op val]
```
</details>

<details><summary><h3>Comments Outside Diff (2)</h3></summary>

1. `packages/engine/src/sql/insight-sql.ts`, line 242-250 ([link](https://github.com/youhaowei/dashframe/blob/664af94cd0462276afe75fcb96df74aae7fa5580/packages/engine/src/sql/insight-sql.ts#L242-L250)) 

   <a href="#">P1</a> **Joined model-mode queries still silently drop filters**

   When an insight has joins and `mode === "model"`, this path returns `SELECT * FROM ${joinedSQL}` with no WHERE clause — filters are still ignored. The PR wires filter clauses into `buildSimpleSQL` and `buildAggregatedSQL`, but this joined-model path was not updated. An insight that has both joins and saved filters will produce unfiltered results whenever it is loaded in model mode, directly contradicting the stated fix.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: packages/engine/src/sql/insight-sql.ts
   Line: 242-250

   Comment:
   **Joined model-mode queries still silently drop filters**

   When an insight has joins and `mode === "model"`, this path returns `SELECT * FROM ${joinedSQL}` with no WHERE clause — filters are still ignored. The PR wires filter clauses into `buildSimpleSQL` and `buildAggregatedSQL`, but this joined-model path was not updated. An insight that has both joins and saved filters will produce unfiltered results whenever it is loaded in model mode, directly contradicting the stated fix.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fengine%2Fsrc%2Fsql%2Finsight-sql.ts%0ALine%3A%20242-250%0A%0AComment%3A%0A**Joined%20model-mode%20queries%20still%20silently%20drop%20filters**%0A%0AWhen%20an%20insight%20has%20joins%20and%20%60mode%20%3D%3D%3D%20%22model%22%60%2C%20this%20path%20returns%20%60SELECT%20*%20FROM%20%24%7BjoinedSQL%7D%60%20with%20no%20WHERE%20clause%20%E2%80%94%20filters%20are%20still%20ignored.%20The%20PR%20wires%20filter%20clauses%20into%20%60buildSimpleSQL%60%20and%20%60buildAggregatedSQL%60%2C%20but%20this%20joined-model%20path%20was%20not%20updated.%20An%20insight%20that%20has%20both%20joins%20and%20saved%20filters%20will%20produce%20unfiltered%20results%20whenever%20it%20is%20loaded%20in%20model%20mode%2C%20directly%20contradicting%20the%20stated%20fix.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youhaowei%2Fdashframe&pr=97&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22youhaowei%2Fdashframe%22%20on%20the%20existing%20branch%20%22yw146-impl%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22yw146-impl%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fengine%2Fsrc%2Fsql%2Finsight-sql.ts%0ALine%3A%20242-250%0A%0AComment%3A%0A**Joined%20model-mode%20queries%20still%20silently%20drop%20filters**%0A%0AWhen%20an%20insight%20has%20joins%20and%20%60mode%20%3D%3D%3D%20%22model%22%60%2C%20this%20path%20returns%20%60SELECT%20*%20FROM%20%24%7BjoinedSQL%7D%60%20with%20no%20WHERE%20clause%20%E2%80%94%20filters%20are%20still%20ignored.%20The%20PR%20wires%20filter%20clauses%20into%20%60buildSimpleSQL%60%20and%20%60buildAggregatedSQL%60%2C%20but%20this%20joined-model%20path%20was%20not%20updated.%20An%20insight%20that%20has%20both%20joins%20and%20saved%20filters%20will%20produce%20unfiltered%20results%20whenever%20it%20is%20loaded%20in%20model%20mode%2C%20directly%20contradicting%20the%20stated%20fix.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youhaowei%2Fdashframe&pr=97&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fengine%2Fsrc%2Fsql%2Finsight-sql.ts%0ALine%3A%20242-250%0A%0AComment%3A%0A**Joined%20model-mode%20queries%20still%20silently%20drop%20filters**%0A%0AWhen%20an%20insight%20has%20joins%20and%20%60mode%20%3D%3D%3D%20%22model%22%60%2C%20this%20path%20returns%20%60SELECT%20*%20FROM%20%24%7BjoinedSQL%7D%60%20with%20no%20WHERE%20clause%20%E2%80%94%20filters%20are%20still%20ignored.%20The%20PR%20wires%20filter%20clauses%20into%20%60buildSimpleSQL%60%20and%20%60buildAggregatedSQL%60%2C%20but%20this%20joined-model%20path%20was%20not%20updated.%20An%20insight%20that%20has%20both%20joins%20and%20saved%20filters%20will%20produce%20unfiltered%20results%20whenever%20it%20is%20loaded%20in%20model%20mode%2C%20directly%20contradicting%20the%20stated%20fix.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=97&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a>

2. `packages/engine/src/sql/insight-sql.ts`, line 241-250 ([link](https://github.com/youhaowei/dashframe/blob/b2e7bf3934050a960e16cbc7694a193c360f67d6/packages/engine/src/sql/insight-sql.ts#L241-L250)) 

   <a href="#">P1</a> **Joined model-mode path still silently drops all filters**

   When an insight has joins and `mode === "model"`, the function returns `SELECT * FROM ${joinedSQL}` with no WHERE clause. `insight.filters` is ignored entirely. Every other code path in this PR now correctly emits filter clauses — `buildSimpleSQL` (model branch), `buildAggregatedSQL` (both the no-config and aggregated branches) — but this joined path was not updated. An insight that has joins and saved filters will produce completely unfiltered results whenever it's loaded in model mode, directly contradicting the PR's stated fix.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: packages/engine/src/sql/insight-sql.ts
   Line: 241-250

   Comment:
   **Joined model-mode path still silently drops all filters**

   When an insight has joins and `mode === "model"`, the function returns `SELECT * FROM ${joinedSQL}` with no WHERE clause. `insight.filters` is ignored entirely. Every other code path in this PR now correctly emits filter clauses — `buildSimpleSQL` (model branch), `buildAggregatedSQL` (both the no-config and aggregated branches) — but this joined path was not updated. An insight that has joins and saved filters will produce completely unfiltered results whenever it's loaded in model mode, directly contradicting the PR's stated fix.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fengine%2Fsrc%2Fsql%2Finsight-sql.ts%0ALine%3A%20241-250%0A%0AComment%3A%0A**Joined%20model-mode%20path%20still%20silently%20drops%20all%20filters**%0A%0AWhen%20an%20insight%20has%20joins%20and%20%60mode%20%3D%3D%3D%20%22model%22%60%2C%20the%20function%20returns%20%60SELECT%20*%20FROM%20%24%7BjoinedSQL%7D%60%20with%20no%20WHERE%20clause.%20%60insight.filters%60%20is%20ignored%20entirely.%20Every%20other%20code%20path%20in%20this%20PR%20now%20correctly%20emits%20filter%20clauses%20%E2%80%94%20%60buildSimpleSQL%60%20%28model%20branch%29%2C%20%60buildAggregatedSQL%60%20%28both%20the%20no-config%20and%20aggregated%20branches%29%20%E2%80%94%20but%20this%20joined%20path%20was%20not%20updated.%20An%20insight%20that%20has%20joins%20and%20saved%20filters%20will%20produce%20completely%20unfiltered%20results%20whenever%20it's%20loaded%20in%20model%20mode%2C%20directly%20contradicting%20the%20PR's%20stated%20fix.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youhaowei%2Fdashframe&pr=97&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22youhaowei%2Fdashframe%22%20on%20the%20existing%20branch%20%22yw146-impl%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22yw146-impl%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fengine%2Fsrc%2Fsql%2Finsight-sql.ts%0ALine%3A%20241-250%0A%0AComment%3A%0A**Joined%20model-mode%20path%20still%20silently%20drops%20all%20filters**%0A%0AWhen%20an%20insight%20has%20joins%20and%20%60mode%20%3D%3D%3D%20%22model%22%60%2C%20the%20function%20returns%20%60SELECT%20*%20FROM%20%24%7BjoinedSQL%7D%60%20with%20no%20WHERE%20clause.%20%60insight.filters%60%20is%20ignored%20entirely.%20Every%20other%20code%20path%20in%20this%20PR%20now%20correctly%20emits%20filter%20clauses%20%E2%80%94%20%60buildSimpleSQL%60%20%28model%20branch%29%2C%20%60buildAggregatedSQL%60%20%28both%20the%20no-config%20and%20aggregated%20branches%29%20%E2%80%94%20but%20this%20joined%20path%20was%20not%20updated.%20An%20insight%20that%20has%20joins%20and%20saved%20filters%20will%20produce%20completely%20unfiltered%20results%20whenever%20it's%20loaded%20in%20model%20mode%2C%20directly%20contradicting%20the%20PR's%20stated%20fix.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=youhaowei%2Fdashframe&pr=97&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a> <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20packages%2Fengine%2Fsrc%2Fsql%2Finsight-sql.ts%0ALine%3A%20241-250%0A%0AComment%3A%0A**Joined%20model-mode%20path%20still%20silently%20drops%20all%20filters**%0A%0AWhen%20an%20insight%20has%20joins%20and%20%60mode%20%3D%3D%3D%20%22model%22%60%2C%20the%20function%20returns%20%60SELECT%20*%20FROM%20%24%7BjoinedSQL%7D%60%20with%20no%20WHERE%20clause.%20%60insight.filters%60%20is%20ignored%20entirely.%20Every%20other%20code%20path%20in%20this%20PR%20now%20correctly%20emits%20filter%20clauses%20%E2%80%94%20%60buildSimpleSQL%60%20%28model%20branch%29%2C%20%60buildAggregatedSQL%60%20%28both%20the%20no-config%20and%20aggregated%20branches%29%20%E2%80%94%20but%20this%20joined%20path%20was%20not%20updated.%20An%20insight%20that%20has%20joins%20and%20saved%20filters%20will%20produce%20completely%20unfiltered%20results%20whenever%20it's%20loaded%20in%20model%20mode%2C%20directly%20contradicting%20the%20PR's%20stated%20fix.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=97&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a>
</details>

<sub>Reviews (4): Last reviewed commit: ["fix(engine): never emit filters referenc..."](https://github.com/youhaowei/dashframe/commit/cae14001db336ed614816d06f1d81bfa57f529c1) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37408192)</sub>

### CODE REVIEW

File: packages/engine/src/sql/insight-sql.ts [hunk 1]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+  for (const fieldId of insight.selectedFields ?? []) {
+    const field = fieldIdMap.get(fieldId);
+    if (field && (field.columnName ?? field.name) === filter.field) {
+      return false; // explicitly a dimension
+    }
+  }
+
+  // Default: treat as dimension (pre-aggregation)
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts:711]: Fixed in 2236d3d. `isMetricFilter` now checks selected-dimension membership **first**, so a column that is both a grouped dimension and a metric source routes to WHERE. Test: "routes a shared dimension+metric column to WHERE".

File: packages/engine/src/sql/insight-sql.ts [hunk 2]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+    }
+    case "in": {
+      const arr = Array.isArray(value) ? value : [value];
+      if (arr.length === 0) return "1=0"; // empty IN → always false
+      return `${columnRef} IN (${arr.map(quoteValue).join(", ")})`;
+    }
+    case "between": {
+      const bv = value as InsightFilterBetweenValue;
+      return `${columnRef} BETWEEN ${quoteValue(bv.low)} AND ${quoteValue(bv.high)}`;
+    }
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts:569]: Fixed in 2236d3d. `between` now guards its value shape — a null/missing-bound value emits an always-true no-op (`1=1`) and warns instead of dereferencing `.low`/`.high`. Test: "emits a no-op predicate (not a throw) for a malformed between value".

File: packages/engine/src/sql/insight-sql.ts [hunk 3]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+      return `${columnRef} LIKE '%${escaped}%' ESCAPE '\\'`;
+    }
+    case "in": {
+      const arr = Array.isArray(value) ? value : [value];
+      if (arr.length === 0) return "1=0"; // empty IN → always false
+      return `${columnRef} IN (${arr.map(quoteValue).join(", ")})`;
+    }
+    case "between": {
+      const bv = value as InsightFilterBetweenValue;
+      return `${columnRef} BETWEEN ${quoteValue(bv.low)} AND ${quoteValue(bv.high)}`;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts:568]: Fixed in 2236d3d — `between` value shape is validated before dereference; malformed → `1=1` + warn, no throw.

File: packages/engine/src/sql/insight-sql.ts [hunk 4]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+    (f) => (f.columnName ?? f.name) === filterField,
+  );
+  if (field) {
+    if (refMode === "alias") {
+      return `"${fieldIdToColumnAlias(field.id)}"`;
+    }
+    return `"${field.columnName ?? field.name}"`;
+  }
+  // Fallback: quote the raw column name
+  return `"${filterField}"`;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts]: Fixed in 2236d3d. Identifier fallbacks now go through `quoteIdentifier`, which doubles embedded `"` per SQL identifier escaping.

File: packages/engine/src/sql/insight-sql.ts [hunk 5]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+    if (metric.columnName === filter.field) return true;
+    if (metricIdToColumnAlias(metric.id) === filter.field) return true;
+  }
+
+  // Check if it matches a selectedField dimension → it's NOT a metric
+  for (const fieldId of insight.selectedFields ?? []) {
+    const field = fieldIdMap.get(fieldId);
+    if (field && (field.columnName ?? field.name) === filter.field) {
+      return false; // explicitly a dimension
+    }
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts]: Fixed in 2236d3d — dimension membership now takes precedence over metric matching in `isMetricFilter`.

File: packages/engine/src/sql/insight-sql.ts [hunk 6]
```diff
@@ -624,13 +895,33 @@ function buildAggregatedSQL(
...
+  // Dimension-vs-metric is derived at compile time from the insight definition:
+  // a filter field that matches a metric column → HAVING (post-aggregation),
+  // everything else → WHERE (pre-aggregation).
+  const hasGroupBy = groupByParts.length > 0;
+  const { whereClause, havingClause } = buildFilterClauses(
+    insight,
+    fieldIdMap,
+    hasGroupBy,
+    "alias",
+  );
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts:963]: Fixed in 2236d3d. The HAVING gate is now `hasMetrics` (query-aggregates), not `hasGroupBy`. A metrics-only insight emits `SUM(...) ... HAVING SUM(...) > n` with no GROUP BY (valid single-group). Test: "routes a metric filter to HAVING even with no GROUP BY".

File: packages/engine/src/sql/insight-sql.ts [hunk 7]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+  refMode: FilterColumnRefMode,
+): { whereClause: string; havingClause: string } {
+  const filters = insight.filters ?? [];
+  if (filters.length === 0) return { whereClause: "", havingClause: "" };
+
+  const wherePredicates: string[] = [];
+  const havingPredicates: string[] = [];
+
+  for (const filter of filters) {
+    const isMetric = hasGroupBy && isMetricFilter(filter, insight, fieldIdMap);
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts]: Fixed in 2236d3d — metric filters route to HAVING whenever the query aggregates (gated on `hasMetrics`), so a metrics-only insight with no GROUP BY puts the predicate on the aggregate in HAVING.

File: packages/engine/src/sql/insight-sql.ts [hunk 8]
```diff
@@ -277,11 +279,25 @@ function buildSimpleSQL(
...
+    const { whereClause } = buildFilterClauses(
+      insight,
+      fieldIdMap,
+      false,
+      "raw",
     );
+
+    let sql = `SELECT ${selectParts.join(", ")} FROM ${tableName} AS "${displayName}"`;
+    if (whereClause) {
+      sql += ` ${whereClause}`;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts:297]: Acknowledged, won't change here. Applying filters in `mode: "model"` is in-scope for this ticket — saved filters were being silently dropped on every path, and model mode emits raw rows where a WHERE is the correct (and only) placement. If filtered preview is undesirable for the Join/Data-Preview surface specifically, that's a call for the caller (`shouldFilter` flag on `BuildInsightSQLOptions`) and belongs in a separate change rather than continuing to drop filters. Flagging rather than silently expanding scope.

File: packages/engine/src/sql/insight-sql.ts [hunk 9]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+ * Everything else defaults to dimension (pre-aggregation WHERE).
+ */
+function isMetricFilter(
+  filter: InsightFilter,
+  insight: Insight,
+  fieldIdMap: Map<string, Field>,
+): boolean {
+  // Check if it matches any metric by column name or metric output alias
+  for (const metric of insight.metrics ?? []) {
+    if (metric.columnName === filter.field) return true;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts]: Fixed in 2236d3d — selected-dimension membership now wins over metric-column matching.

File: packages/engine/src/sql/insight-sql.ts [hunk 10]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+  columnRef: string,
+  filter: InsightFilter,
+): string {
+  const { operator, value } = filter;
+
+  switch (operator) {
+    case "eq":
+      return `${columnRef} = ${quoteValue(value)}`;
+    case "ne":
+      return `${columnRef} <> ${quoteValue(value)}`;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts:527]: Fixed in 2236d3d — `eq`/`ne` with null/undefined now emit `IS NULL`/`IS NOT NULL`. Tests added for both.

File: packages/engine/src/sql/insight-sql.ts [hunk 11]
```diff
@@ -459,6 +475,253 @@ function processSingleJoin(
...
+  filterField: string,
+  fieldIdMap: Map<string, Field>,
+  refMode: FilterColumnRefMode,
+): string {
+  const field = Array.from(fieldIdMap.values()).find(
+    (f) => (f.columnName ?? f.name) === filterField,
+  );
+  if (field) {
+    if (refMode === "alias") {
+      return `"${fieldIdToColumnAlias(field.id)}"`;
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts]: Partially addressed: the fallback now escapes via `quoteIdentifier`. The deeper join-key case (a filter targeting a dropped join-key alias) is a real edge but out of scope for this correctness fix — filters here resolve against the field set, and a filter on a dropped duplicate join key is a pre-existing modeling gap, not introduced by this PR. Flagging for a follow-up rather than expanding scope.

File: packages/engine/src/sql/insight-sql.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts:297]: Fixed in cae1400 — you were right, this was a real consistency defect (single-table model previews filtered, joined model path didn't). Resolved by gating the filter clause to QUERY mode only: model/preview SQL is now never filtered on either path. A preview shows the raw source rows so the user can see what they're working with before filtering. Tested: model-mode SQL has no filter clause even with saved filters; query-mode does (and runtime-verified against DuckDB — model preview returns all rows, query returns the filtered set).

File: packages/engine/src/sql/insight-sql.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/engine/src/sql/insight-sql.ts]: Fixed in cae1400 — agreed this was a runtime-crash bug my filter code introduced (silent-drop became a broken column ref). `processSingleJoin` drops the right join-key from the joined subquery; `buildJoinedSQL` now returns the exact column set actually present (dropped keys excluded), and `resolveFilterColumnRef` returns null for any filter field not in that set. `buildFilterClauses` then skips-and-warns (field name only, never the value) rather than emitting `WHERE "<missing>" = …`. A filter on the join column still resolves to the retained left key, so it keeps working. Runtime-verified against DuckDB: the join-key filter and a stale/ghost filter both EXECUTE now instead of crashing; retained joined-table columns still filter correctly.
